contracts/src/modules/Coverage.sol

```diff
@@ -6,81 +6,105 @@ import {DataTypes} from "../libraries/DataTypes.sol";
-    ) ModuleBase(registrar, _fees) {
+        address _usdc,
+        uint256 _coveragePeriod,  // In seconds
+        uint256 _reservesLockupPeriod,  // In seconds
+        uint256 _reserveRatio
+    ) ModuleBase(registrar, _fees) ERC20("DenotaCoverageToken", "DCT"){
         _URI = __baseURI;
         usdc = _usdc;
-    }
-
-    // Admin can add an address to the whitelist
-    function addToWhitelist(address _address) external onlyOwner {
-        isWhitelisted[_address] = true;
-    }
-
-    // Admin can remove an address from the whitelist
-    function removeFromWhitelist(address _address) external onlyOwner {
-        isWhitelisted[_address] = false;
-    }
-
-    function fundPool(uint256 fundingAmount) public {
-        IERC20(usdc).safeTransferFrom(_msgSender(), address(this), fundingAmount);
-
-        // LPs receive pool tokens in return
-        // TODO: figure out token issuance and redemption
+        coveragePeriod = _coveragePeriod;
+        reservesLockupPeriod = _reservesLockupPeriod;
+        reserveRatio = _reserveRatio;
     }
 
     function processWrite(
         address caller,
         address _owner,
-        uint256 cheqId,
+        uint256 notaId,
         address currency,
         uint256 /*escrowed*/,
         uint256 instant,
         bytes calldata initData
     ) public override onlyRegistrar returns (uint256) {
         require(isWhitelisted[caller], "Not whitelisted");
+        require(_owner == address(this), "Risk fee not paid to pool"); // TODO registrar assumes that the owner is the one being paid
+        require(currency == usdc, "Incorrect currency");
+
+        uint256 maturityDate = block.timestamp + coveragePeriod;
+
+        if (poolStart == 0) {
+            poolStart = block.timestamp;
+            reservesReleaseDate = block.timestamp + reservesLockupPeriod;
+        } else {
+            require(maturityDate <= reservesReleaseDate);
+        }
 
         (address holder, uint256 amount, uint256 riskScore) = abi.decode(
             initData,
             (address, uint256, uint256)
         );
 
-        require(instant == (amount / 10000) * riskScore, "Risk fee not paid");
-        // TODO: maybe onramp should own nota? (currently the registrar assumes that the owner is the one being paid)
-        require(_owner == address(this), "Risk fee not paid to pool");
-        require(currency == usdc, "Incorrect currency");
+        require(instant == (amount / 10_000) * riskScore, "Risk fee not paid");
+
+        uint256 currentCoverage = (totalReserves - availableReserves);
+        uint256 newCoverage = currentCoverage + amount;
+
+        uint256 newReserveBPS = (totalReserves / newCoverage) * 100;  // 1_000 / 800 => 800 * 100 => 8_000
+        if (newReserveBPS < reserveRatio) revert("Exceeds reserve ratio");
+
+        uint256 scaledAmount = (amount * reserveRatio) / 10_000;
+        availableReserves -= scaledAmount; // scale down by reserve ratio (amount=100 => amount=5 if rR=2_000)
```

Comment:
hmmmm why not just do 
```
availableReserves -= amount * reserveRatio
if (availableReserves < 0) revert("Exceeds reserve ratio");
```

Reply:
The ratio is in BPS since decimals aren't well supported